keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,297 @@
+2. If kubelet sets `--enforce-node-allocatable=system-reserved`, `--system-reserved=[a]` and `--system-reserved-cgroup=[b]`, we set `memory.min=[a]` for node level cgroup `[b]`
+3. If kubelet sets `--enforce-node-allocatable=pods`, we set `memory.min=sum(pod.spec.containers[i].resources.requests[memory])` for kubepods cgroup
+
+### Interactive
+New `Unified` field will be added in both CRI and QoS Manager for cgroups v2 extra parameters. It is recommended to has same semantics with opencontainers/runtime-spec#1040
+
+- container level: `Unified` added in `LinuxContainerResources`
+- pod/node level: `Unified` added in `cm.ResourceConfig` 
+
+### Workflow
+#### Container Level 
+![](./container-level.png)
+
+#### Pod Level 
+![](./pod-level.png)
+
+#### Node Level
+![](./node-level.png)
+
+### Cgroup Hierarchy 
+Container level:
+```
+/cgroup2/kubepods/pod<UID>/<container-id>/memory.min=pod.spec.containers[i].resources.requests[memory]
+/cgroup2/kubepods/pod<UID>/<container-id>/memory.high=85%*pod.spec.containers[i].resources.limits[memory] // Burstable
+```
+
+Pod level:
+```
+/cgroup2/kubepods/pod<UID>/memory.min=sum(pod.spec.containers.resources.requests[memory])
+```
+
+Node level:
+```
+/cgroup2/kubepods/memory.min=sum(pod.spec.containers[i].resources.requests[memory])
+/cgroup2/<kube-reserved-cgroup,system-reserved-cgroup>/memory.min=<kube-reserved,system-reserved>
+```
+
+### Cgroup v2 Support
+After Kubernetes v1.19, kubelet can identify cgroups v2 and do the convention. Since [v1.0.0-rc93](https://github.com/opencontainers/runc/releases/tag/v1.0.0-rc93), runc supports `Unified` to pass through cgroups v2 parameters. So we use this variable to pass `memory.min` when cgroups v2 mode is detected.
+
+### Container Runtime Interface (CRI) Changes
+We need add new field `Unified` in CRI api which is basically passthrough for OCI spec Unified field and has same semantics: opencontainers/runtime-spec#1040
+
+```
+type LinuxContainerResources struct {
+    ...
+    Unified map[string]string `json:"unified,omitempty"
+}
+```
+
+### Test Plan
+For `Alpha`, unit tests were added to test functionality for container/pod/node level cgroup with containerd and CRI-O.
+
+For `Beta`, e2e-node tests were added, tests were added for docker
+
+For `GA`, the introduced e2e tests will be promoted to conformance. It was also verified that all e2e coverage was proper and CRI's had tests in their respective repos testing this feature.
+
+
+### Graduation Criteria
+
+#### Alpha Graduation
+- [cgroup_v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2) is in `Alpha`
+- Memory QoS is implemented for new feature gate
+- Memory QoS is covered by proper tests
+- Memory QoS supports containerd, cri-o
+
+#### Beta Graduation
+- [cgroup_v2](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2254-cgroup-v2) is in `Beta`
+- Memory QoS is covered by unit and e2e-node tests
+- Memory QoS supports containerd, cri-o and docker
```

Comment:
What is meant with `docker` here? Do you mean dockershim or have docker made a new CRI I am not aware of?

Reply:
Yes, I have fixed the typo.